sound/soc/sof/intel/hda-trace.c

```diff
@@ -90,5 +90,22 @@ int hda_dsp_trace_trigger(struct snd_sof_dev *sdev, int cmd)
 {
 	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
 
+	if (!sdev->dtrace_is_supported)
+		return 0;
+
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_START:
+		/* FIXME: return 0 when dtrace_is_enabled? */
+		if (sdev->dtrace_is_enabled || !sdev->dma_trace_pages)
+			return -EINVAL;
+		break;
+	case SNDRV_PCM_TRIGGER_STOP:
+		if (!sdev->dtrace_is_enabled)
+			return 0;
+		break;
+	default:
+		break;
+	}
+
```

Comment:
I think this logic has to be in trace.c and not in Intel specific code.

We also have to be careful with how we use dtrace_is_enabled. If we start starting/stopping trace based on PM transitions, we may have to start tracking new state (dtrace_is_running perhaps). Otherwise we might end up in problems like:

   - dtrace is enabled -> dtrace_is_enabled
   - D0ix is entered -> trigger stop -> dtrace_is_enabled = false 
   - sof-logger is terminated -> snd_sof_release_trace() is called bu as dtrace_is_enabled is not set, cleanup is not done properly

This might not happen in practise, but there are a lot of place in trace.c where dtrace_is_enable is linked to whether sof-logger has opened the trace node or not. Trigger start/stop is clearly a substate of dtrace_is_enable==true, so at least all uses of dtrace_is_enabled need to be carefully checked.

Reply:
> I think this logic has to be in trace.c and not in Intel specific code.

@kv2019i OK. What about we put such code in snd_sof_dma_trace_trigger() in ops.h?

> we may have to start tracking new state (dtrace_is_running perhaps)

@kv2019i I'm not sure I understand your proposal correctly or not. Maybe you mention sdev->dtrace_is_supported here? If we enable dma trace feature, sdev->dtrace_is_supported will be set. If we trigger start dma trace, sdev->dtrace_is_enabled will be set.

> dtrace is enabled -> dtrace_is_enabled

This situation is handled in this patch "case SNDRV_PCM_TRIGGER_START:" and it will check sdev->dtrace_is_enabled. What I'm not sure is whether we should return 0 or return -EINVAL in this case. In my patch, I returned -EINVAL.

> D0ix is entered -> trigger stop -> dtrace_is_enabled = false

Yes, we set dtrace_is_enabled in the caller function each time.

> sof-logger is terminated -> snd_sof_release_trace() is called bu as dtrace_is_enabled is not set, cleanup is not done properly

Actually we will always set dtrace_is_enabled whenever dma trace is trigger started. This is well handled in the caller.

Reply:
@libinyang  Hmm, I still think there is potential confusing. We have three states that are tracked:
  - is trace supported        
       - -> tracked with *dtrace_is_supported*, set via kernel module param
  - is trace initialized, pages alloced
       - -> tracked with *dtrace_is_enabled*, initially set at driver probe
  - is trace DMA running
       - your patch uses *dtrace_is_enabled* to track this as well, but I was wondering if we get into situations where we don't do a proper cleanup if DMA trace is not running (trigger_stop has been run) but it is still initialized (pages allocated)

Before your patch, there was not problem as DMA trigger was only done in snd_sof_init_trace_ipc() so the states were always aligned. But now with separate trigger stop/starts, situation gets more complicated.

Just using dtrace_is_enabled might still be safe, but then we just need to double-check that when driver is released, we clean the resources in all cases.

Reply:
> is trace initialized, pages alloced
> -> tracked with dtrace_is_enabled, initially set at driver probe

@kv2019i dtrace_is_enabled is set after the dma trace is triggered start. And every time we trigger stop the dma trace, we will set dtrace_is_enabled to 0. To check "pages alloced", we will check sdev->dma_trace_pages. As long as we set dtrace_is_enabled correctly every time when we trigger start/stop dma trace, we can use dtrace_is_enable for trace DMA is running or not.

The below scenarios are what I can figure:

For module reload
sof_pci_remove() -> dma trace will be triggered stop and dtrace_is_enabled will be set 0
so next time sof reload, dma trace will be initialized successfully.

For suspend/resume or runtime suspend/resume
sof_suspend() will be called and so snd_sof_release_trace(sdev) will be called. 
So dma trace will be triggered stop and dtrace_is_enabled will be set 0
Next time resume, snd_sof_init_trace_ipc() will be called and dma trace will be initialized successfully

For S0ix
In the suspend, this patch will trigger stop dma trace and dtrace_is_enabled will be set 0
When resume from s0ix, this patch will trigger start dma trace and dtrace_is_enabled will be set 1